### DISCUSSION TITLE
how can you make on_enter in ExtensionResultItem run multiple actions

### DISCUSSION THREAD

**Discussion by decoherencer** [Ulauncher/Ulauncher#1208] [Extensions and Themes]
I have an url that can be copied to clipboard from rendered results with `CopyToClipboardAction` upon pressing enter.
but, what if I have a function that does something on URL and now I want to run both this function and also copy the URL to clipboard upon enter.
I saw there is ExtensionCustomAction in https://docs.ulauncher.io/en/5.15.1/extensions/actions.html#extensioncustomaction that could help to combine both, but is there simpler way or code example to do it?
Thanks!

**Comment by decoherencer** [comment]
One more question each item in rendered results list has a dynamic icon that changes for each event trigger, 
but ulauncher is keeping a static icon that it first encounters and is not changing it for future triggers.
Is this a bug or design decision to cache icons?

**Reply by friday** [reply]
It should only cache the same input giving the same output. So if you for example have a path to an image and overwrite that image it will not get the new one, but if you use a different path it should.

**Comment by decoherencer** [comment]
I noticed this https://github.com/Ulauncher/Ulauncher/discussions/966 but still not sure how to call custom function with alt_enter

**Reply by friday** [reply]
alt_enter is if you want to run a different action (ex "open in browser" for enter and "open in incognito browser" for alt_enter.

If you want to run two Ulauncher actions from the same call, then you need ActionList

**Comment by friday** [comment]
You need to use a custom action with [actionlist](http://docs.ulauncher.io/en/stable/extensions/actions.html#actionlist) now. You can see how other extensions used it here: https://github.com/search?q=ulauncher.api.shared.action.ActionList&type=code

This is one of the major things I don't like about Ulauncher now and want to change for v6, but I have too little time to work on Ulauncher now.

**Reply by decoherencer** [reply]
I am still not sure of how to call a simple python function declared in some file, such as `example(args)` from the action list :sweat: or even replacing on_enter with just that single function call

**Reply by friday** [reply]
In that case I *think* (unable to check right now and memory isn't fresh) you can use http://docs.ulauncher.io/en/stable/extensions/actions.html#extensioncustomaction and handle your function through that and do all the non-action code, then return the action/actionlist you want to run after that.

The actions mostly exist because they have a call method that needs to run in the app rather than the extension runtime. But some actions can actually run in the extension runtime, or be substituted with custom python code.

**Reply by decoherencer** [reply]
Thank you! I got it working it after looking at example usage here - https://github.com/Dankni95/backup/blob/46db1f4188fb1133ee721e801bd85195aa9c6704/extensions/com.github.dcervenkov.ulauncher-z-search/main.py
